src/Search/Search.Tests/ModelComparer/ModelComparer.cs

```diff
@@ -0,0 +1,220 @@
+// license information.
+
+namespace Microsoft.Azure.Search.Tests.Utilities
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+
+    /// <summary>
+    /// Compares instances of a type for structural equality, taking into account public properties, collections, and dictionaries.
+    /// </summary>
+    /// <typeparam name="T">Type of objects to compare.</typeparam>
+    /// <remarks>
+    /// <para>
+    /// Supported types include enums, primitives, nullables, List, Dictionary, and DTOs having public properties
+    /// composed of the supported types (including other nested DTOs). The rules are different for different types.
+    /// </para>
+    /// <para>
+    /// Any type implementing IEnumerable, including dictionaries, are compared element by element and compose with the other mapping
+    /// rules on element type.
+    /// </para>
+    /// <para>
+    /// DTOs are compared by the values of their public properties. Note that comparisons are performed according to the actual most-derived
+    /// type of the objects (which must match), not the apparent static type of the references.
+    /// </para>
+    /// <para>
+    /// Enums, primitive types, and any other type not falling into the other categories are compared using Object.Equals().
+    /// </para>
+    /// </remarks>
+    public class ModelComparer<T> : IEqualityComparer<T>
+    {
+        private IEqualityComparer _comparer = new DynamicModelComparer(typeof(T));
+
+        public bool Equals(T x, T y) => _comparer.Equals(x, y);
+
+        public int GetHashCode(T obj) => obj.GetHashCode();
+
+        private class DynamicModelComparer : IEqualityComparer
+        {
+            private Type _type;
+
+            public DynamicModelComparer(Type type)
+            {
+                _type = type;
+            }
+
+            int IEqualityComparer.GetHashCode(object obj) => obj.GetHashCode();
+
+            bool IEqualityComparer.Equals(object x, object y)
+            {
+                if (_type.CanBeNull() && x == null || y == null)
+                {
+                    return CompareNull(x, y);
+                }
+
+                // At this point x and y are guaranteed to be non-null (possibly because they are boxed value types).
+
+                if (_type.ImplementsGenericEquatable())
+                {
+                    return CompareEquatables(x, y);
+                }
+
+                Type enumerable = _type.GetIEnumerable();
+                if (enumerable != null)
+                {
+                    return CompareEnumerables(enumerable, x, y);
+                }
+
+                Type actualType = x.GetType();
+                if (y.GetType() != actualType)
+                {
+                    // The only case where x and y can have different types and still be equal is integer comparison.
```

Comment:
It doesn't happen in practice because our API types everything as `double`. The reason the integer case happens frequently is that things are often typed as `int`, but end up deserializing as `long` because that is the default integer type used by JSON.NET.